Jump started the statement part of the grammar

Doryanni opened the pull request:
#7 Added block, while, assignment, array assignment and if statements.

Commit:
Jump started the statement part of the grammar

Doryanni commented:
Making some changes currently as per the post in #14

Commit:
Minor Cleanup

Doryanni committed:
Merge branch 'master' into grammar-statements

baziotis commented on grammar.ebnf:
```diff
@@ -20,8 +20,19 @@ StructDecl = "struct", Identifier, "{", { StructField }, "}";
 
 
+Statement = "{", { Statement }, "}"
+          | Type, Identifier, ";"
```
What's this?

Doryanni replied:
It's a block statement. It opens up a new scope and can be used in conjunction with if and while statements. Although now I see that my code in while loops obligates braces whilst the if does not.

I'll also make it so that the variable definition works with multiple variables per type.

Doryanni replied:
I pushed a new commit making the if and while consistent. If you believe it's better to have mandatory braces, I'll add them.

baziotis replied:
> It's a block statement.

I'm referring to line 24. It seems like a variable _declaration_. Which is... a declaration, not a statement :). We may allow "declare anywhere" but for the time being, we'll just have all variable declarations at the start.

Doryanni replied:
Oh fair enough. You marked both lines so I assumed you meant both.

baziotis replied:
Actually I didn't and as I said [here](https://github.com/ACM-UoA-Student-Chapter/dil/pull/16#discussion_r352233741) I don't know of a way to make Github mark only one specific line.

baziotis commented on grammar.ebnf:
```diff
@@ -20,8 +20,19 @@ StructDecl = "struct", Identifier, "{", { StructField }, "}";
+          | "while", "(", Expression, ")", "{", { Statement }, "}"
+          | Lval, "=", Expression, ";"
+          | Lval, "[", Expression, "]", "=", Expression, ";"
+          | "if", "(", Expression, ")", Statement;
```
Please make curly braces obligatory.

Doryanni replied:
Is it really necessary? One-liner ifs are quite common and putting obligatory braces seems a bit excessive and makes the code a bit harder to read, especially for people new to programming, who are this language's target audience (even if it's a mock project :p )

baziotis replied:
Well, for me it makes the code harder to read to _not_ put obligatory braces (not in all cases, but in most in my experience).
However it's true that most people seem to like it so I agree, we'll keep it your way.

baziotis commented on grammar.ebnf:
```diff
@@ -20,8 +20,19 @@ StructDecl = "struct", Identifier, "{", { StructField }, "}";
+          | "if", "(", Expression, ")", Statement;
+
+
+Lval = Identifier, { ".", Identifier };
```
This is a good thought but I propose that we restrict that to just `Identifier` in the start. Then we can start doing more complicated things.

Doryanni replied:
Fair enough.

baziotis commented:
Please write a PR description that makes clear what parts of the grammar this PR handles.

Commit:
Support for multiple variable definitions of a single type in a line. Also removed while brace

Commit:
Minor repetition fixes for while and if

baziotis commented:
Please use consistently the final `;` in the rules.
Other than that, we need `return` statements and function calls and I think we're done.
The latter is subtle because a function call is also an expression. Languages like C / C++
couple statements and expressions so that an expression can be used as a statement. For example:
```
Statement = ...
          | Expression
```
This is not the best practice so for now, we can just make the rule for function calls and make
only that be able to be a statement.

Doryanni commented:
> Please use consistently the final `;` in the rules.

To what do you refer to?

baziotis commented:
> > Please use consistently the final `;` in the rules.
> 
> To what do you refer to?

Nothing, my bad. Me trying to get used to EBNF sometimes complicates things :)

Commit:
function call and return added to statements

Commit:
function call added to statements

Commit:
Merge branch 'grammar-statements' of https://github.com/wilhelmstein/dil into grammar-statements

Doryanni commented:
I tried changing my commit message, that is why multiple commits appeared.

spy-t commented:
@WilhelmStein take a look at [amend](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---amend)

baziotis commented:
Also in [squash](https://www.devroom.io/2011/07/05/git-squash-your-latests-commits-into-one/).
Experiment if you like, this is why we're making this mock project in the first place.

baziotis commented on grammar.ebnf:
```diff
@@ -20,8 +20,19 @@ StructDecl = "struct", Identifier, "{", { StructField }, "}";
+          | Identifier, "(", FuncCallArgList, ")", ";" 
+          | "return", Expression, ";";
+
+FuncCallArgList = [Identifier, { ",", Identifier }];
```
This is unfortunately not correct. Say you write this call: `a(1, 2)`.

Doryanni replied:
Oops, silly mistake. Will refrain from posting so eagerly :p

Commit:
minor fix in function call arguments

baziotis commented on grammar.ebnf:
```diff
@@ -20,8 +20,19 @@ StructDecl = "struct", Identifier, "{", { StructField }, "}";
+          | Identifier, "(", FuncCallArgList, ")", ";" 
+          | "return", Expression, ";";
+
+FuncCallArgList = [Expression, { ",", Expression }];
```
Well done!

baziotis commented on grammar.ebnf:
```diff
@@ -20,8 +20,19 @@ StructDecl = "struct", Identifier, "{", { StructField }, "}";
+          | Identifier, "=", Expression, ";"
+          | Identifier, "[", Expression, "]", "=", Expression, ";"
+          | "if", "(", Expression, ")", Statement
+          | Identifier, "(", FuncCallArgList, ")", ";" 
```
For educational purposes in case you don't have it in your mind, in languages like C / C++, the `Identifier` here should be an expression, because in `p->a(1, 2)`, `a` might be a function pointer. And a function pointer can result from arbitrary expressions. In C++ it can result from method referencing etc.
Here we don't have function pointers or methods (i.e. member functions) so no need for that.

Doryanni replied:
Yes, that is why I did not make it an expression in the first place. Thanks for the note nonetheless.

baziotis commented on grammar.ebnf:
```diff
@@ -20,8 +20,19 @@ StructDecl = "struct", Identifier, "{", { StructField }, "}";
+          | Identifier, "[", Expression, "]", "=", Expression, ";"
+          | "if", "(", Expression, ")", Statement
+          | Identifier, "(", FuncCallArgList, ")", ";" 
+          | "return", Expression, ";";
```
Just make the `Expression` optional since we can have functions that return nothing.

Commit:
made expression optional

baziotis committed:
Added assignment for fields.

Note that we only need to support Identifier.Identifier and not an arbitrary number of .field, like a.b.c = 2 or a.b.c.d = 2;
That is because structs and arrays are always references, i.e. pointers. So, the above can be achieved as:
B b = a.b;  // This is a pointer - any change to this will be reflected to a.b
b.c = 2;
For the same reason, we don't need to support something like:
a.b[2] = 3;
because we can achieve that as:
Arr arr = a.b;  // arr is a pointer under the hood - any change to its elements will be reflected to a.b
arr = a.b;

baziotis commented:
I added an important commit. It was a little subtle and I hadn't anticipated it so I thought better do it myself and add an explanation.
I think we're ok, thanks again. Merging!

baziotis commented:
There were some problems when this got merged with Git. I tried to solve it by squashing into one.
